File: Data Structure/Stack/acmicpc.net/17299/Main.py
# import sys
# input = sys.stdin.readline
file = open("./01 ing/acmicpc.net/sample.txt", "r")
input = file.readline
n = int(input())
arr = list(map(int, input().split(" ")))

f = [0]*1000001;
s = [0];
ans = [0] * n;
for i in range(n):
    f[arr[i]] +=1

for i in range(1, n):
    while s and f[arr[s[-1]]] <f[arr[i]]:
        ans[s.pop()] = arr[i]
    s.append(i);

while s:
    ans[s.pop()] = -1
print(" ".join(map(str, ans)))

File: Data Structure/Stack/acmicpc.net/17299/Main.js
"use strict";

const inputFilePath = "/sample.txt";
const input = require("fs")
  .readFileSync(__dirname + inputFilePath)
  .toString()
  .trim()
  .split("\n");
// const input = require("fs").readFileSync("/dev/stdin").toString().trim().split("\n");

const n = Number(input[0]);
const arr = input[1].split(" ").map((x) => Number(x));
const s = [0];
const f = new Array(1000001).fill(0);
const ans = [];
for (let i = 0; i < n; i++) {
  f[arr[i]]++;
}

for (let i = 1; i < n; i++) {
  while (s.length && f[arr[s[s.length - 1]]] < f[arr[i]]) {
    ans[s.pop()] = arr[i];
  }
  s.push(i);
}
while (s.length) {
  ans[s.pop()] = -1;
}
console.log(ans.join(" "));
